README.md
# Image Card

Customize a card with image by 2 type: fill or transparent

<img src="https://raw.githubusercontent.com/zeref278/image_card/main/example/demo.gif"/>

## Usage
### Fill image card

<img src="https://raw.githubusercontent.com/zeref278/image_card/main/example/screenshot.png" width="300"/>

```dart
FillImageCard(
  width: 200,
  heightImage: 140,
  imageProvider: AssetImage('assets/mockup.png'),
  tags: [_tag('Category', () {}), _tag('Product', () {})],
  title: _title(),
  description: _content(),
),
```

### Transparent image card

<img src="https://raw.githubusercontent.com/zeref278/image_card/main/example/screenshot1.png" width="300"/>

```dart
TransparentImageCard(
  width: 200,
  imageProvider: AssetImage('assets/mockup.png'),
  tags: [ _tag('Product', () {}), ],
  title: _title(color: Colors.white),
  description: _content(color: Colors.white),
),
```

## Parameter

You can see parameter detail in github repo.


CHANGELOG.md
# Releases
# 0.0.1
- Initial open source release.
- Usage and example.
# 0.0.2
- Update README.
# 0.0.3
- Update README.
# 0.0.4
- Dart formatted.

lib/src/fill_image_card.dart
import 'package:flutter/material.dart';
import 'package:image_card/src/image_card_content.dart';

class FillImageCard extends StatelessWidget {
  const FillImageCard({
    Key? key,
    this.width,
    this.height,
    this.heightImage,
    this.borderRadius = 6,
    this.contentPadding,
    required this.imageProvider,
    this.tags,
    this.title,
    this.description,
    this.footer,
    this.color = Colors.white,
    this.tagSpacing,
    this.tagRunSpacing,
  }) : super(key: key);

  /// card width
  final double? width;

  /// card height
  final double? height;

  /// image height
  final double? heightImage;

  /// border radius value
  final double borderRadius;

  /// spacing between tag
  final double? tagSpacing;

  /// run spacing between line tag
  final double? tagRunSpacing;

  /// content padding
  final EdgeInsetsGeometry? contentPadding;

  /// image provider
  final ImageProvider imageProvider;

  /// list of widgets
  final List<Widget>? tags;

  /// card color
  final Color color;

  /// widget title of card
  final Widget? title;

  /// widget description of card
  final Widget? description;

  /// widget footer of card
  final Widget? footer;

  @override
  Widget build(BuildContext context) {
    return Container(
      width: width,
      height: height,
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(borderRadius),
        color: color,
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisSize: MainAxisSize.min,
        children: [
          ClipRRect(
            borderRadius: BorderRadius.only(
              topLeft: Radius.circular(borderRadius),
              topRight: Radius.circular(borderRadius),
            ),
            child: Image(
              image: imageProvider,
              width: width,
              height: heightImage,
              fit: BoxFit.cover,
            ),
          ),
          ImageCardContent(
            contentPadding: contentPadding,
            tags: tags,
            title: title,
            footer: footer,
            description: description,
            tagSpacing: tagSpacing,
            tagRunSpacing: tagRunSpacing,
          ),
        ],
      ),
    );
  }
}


lib/src/image_card_content.dart
import 'package:flutter/material.dart';

class ImageCardContent extends StatelessWidget {
  const ImageCardContent({
    Key? key,
    this.contentPadding,
    this.tags,
    this.title,
    this.description,
    this.footer,
    this.tagSpacing,
    this.tagRunSpacing,
  }) : super(key: key);

  final EdgeInsetsGeometry? contentPadding;
  final List<Widget>? tags;
  final double? tagSpacing;
  final double? tagRunSpacing;

  final Widget? title;
  final Widget? description;
  final Widget? footer;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: contentPadding ??
          const EdgeInsets.only(top: 8, bottom: 12, left: 8, right: 8),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisSize: MainAxisSize.min,
        children: [
          if (tags != null)
            Wrap(
              spacing: tagSpacing ?? 12,
              runSpacing: tagRunSpacing ?? 10,
              children: tags!,
            ),
          if (title != null || description != null)
            Padding(
              padding: const EdgeInsets.only(top: 8.0),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  if (title != null) title!,
                  if (title != null && description != null)
                    const SizedBox(
                      height: 2,
                    ),
                  if (description != null) description!,
                ],
              ),
            ),
          if (footer != null)
            Padding(
              padding: const EdgeInsets.only(top: 2),
              child: footer!,
            ),
        ],
      ),
    );
  }
}


lib/src/transparent_image_card.dart
import 'package:flutter/material.dart';

import 'image_card_content.dart';

class TransparentImageCard extends StatelessWidget {
  const TransparentImageCard({
    Key? key,
    this.width,
    this.height,
    this.contentMarginTop,
    this.borderRadius = 6,
    this.contentPadding,
    required this.imageProvider,
    this.tags,
    this.title,
    this.description,
    this.footer,
    this.startColor,
    this.endColor,
    this.tagSpacing,
    this.tagRunSpacing,
  }) : super(key: key);

  /// card width
  final double? width;

  /// card height
  final double? height;

  /// padding from top of card to content
  final double? contentMarginTop;

  /// border radius value
  final double borderRadius;

  /// spacing between tag
  final double? tagSpacing;

  /// run spacing between line tag
  final double? tagRunSpacing;

  /// content padding
  final EdgeInsetsGeometry? contentPadding;

  /// image provider
  final ImageProvider imageProvider;

  /// list of widgets
  final List<Widget>? tags;

  /// color gradient start, default [0xff575757] with opacity 0
  final Color? startColor;

  /// color gradient end, default [0xff000000]
  final Color? endColor;

  /// widget title of card
  final Widget? title;

  /// widget description of card
  final Widget? description;

  /// widget footer of card
  final Widget? footer;

  @override
  Widget build(BuildContext context) {
    final Widget content = ImageCardContent(
      contentPadding: contentPadding,
      tags: tags,
      title: title,
      footer: footer,
      description: description,
      tagSpacing: tagSpacing,
      tagRunSpacing: tagRunSpacing,
    );

    return _buildBody(content);
  }

  Widget _buildBody(Widget content) {
    return Stack(
      children: [
        ClipRRect(
          borderRadius: BorderRadius.circular(borderRadius),
          child: ShaderMask(
            shaderCallback: (bound) {
              return LinearGradient(
                begin: Alignment.topCenter,
                end: Alignment.bottomCenter,
                colors: <Color>[
                  startColor ?? const Color(0xff575757).withOpacity(0),
                  endColor ?? const Color(0xff000000),
                ],
                //tileMode: TileMode.,
              ).createShader(bound);
            },
            blendMode: BlendMode.srcOver,
            child: Container(
              width: width,
              height: height,
              decoration: BoxDecoration(
                borderRadius: BorderRadius.circular(borderRadius),
                image: DecorationImage(
                  image: imageProvider,
                  fit: BoxFit.cover,
                ),
                //tileMode: TileMode.,
              ),
              padding: EdgeInsets.only(top: contentMarginTop ?? 100),
              child: content,
            ),
          ),
        ),
        Positioned(
          child: Container(
            width: width,
            height: height,
            decoration: BoxDecoration(
              borderRadius: BorderRadius.circular(borderRadius),
              color: Colors.transparent,
            ),
            padding: EdgeInsets.only(top: contentMarginTop ?? 100),
            child: content,
          ),
        ),
      ],
    );
  }
}


example/lib/image_card_demo_page.dart
import 'package:flutter/material.dart';
import 'package:image_card/image_card.dart';

class ImageCardDemoPage extends StatelessWidget {
  const ImageCardDemoPage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Color(0xFFE5E5E5),
      appBar: AppBar(
        title: Text('Image card demo'),
      ),
      body: SafeArea(
        child: SingleChildScrollView(
            padding: const EdgeInsets.symmetric(horizontal: 24, vertical: 16),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              mainAxisSize: MainAxisSize.min,
              children: [
                const SizedBox(height: 12),
                SingleChildScrollView(
                  scrollDirection: Axis.horizontal,
                  child: Row(
                    children: [
                      FillImageCard(
                        width: 200,
                        heightImage: 140,
                        imageProvider: AssetImage('assets/mockup.png'),
                        tags: [_tag('Category', () {}), _tag('Product', () {})],
                        title: _title(),
                        description: _content(),
                      ),
                      const SizedBox(width: 12),
                      FillImageCard(
                        width: 200,
                        heightImage: 140,
                        imageProvider: AssetImage('assets/mockup.png'),
                        tags: [_tag('Category', () {}), _tag('Product', () {})],
                        title: _title(),
                        description: _content(),
                      ),
                      const SizedBox(width: 12),
                      FillImageCard(
                        width: 200,
                        heightImage: 140,
                        imageProvider: AssetImage('assets/mockup.png'),
                        description: _content(),
                      ),
                      const SizedBox(width: 12),
                      FillImageCard(
                        width: 200,
                        heightImage: 140,
                        imageProvider: AssetImage('assets/mockup.png'),
                        tags: [_tag('Category', () {})],
                      ),
                    ],
                  ),
                ),
                const SizedBox(height: 12),
                SingleChildScrollView(
                  scrollDirection: Axis.horizontal,
                  child: Row(
                    children: [
                      FillImageCard(
                        width: 200,
                        heightImage: 140,
                        imageProvider: AssetImage('assets/mockup.png'),
                        tags: [_tag('Category', () {})],
                        title: _title(),
                        description: _content(),
                        footer: _footer(),
                      ),
                      const SizedBox(width: 12),
                      FillImageCard(
                        width: 200,
                        heightImage: 140,
                        imageProvider: AssetImage('assets/mockup.png'),
                        tags: [_tag('Category', () {})],
                        title: _title(),
                        description: _content(),
                        footer: _footer(),
                      ),
                      const SizedBox(width: 12),
                      FillImageCard(
                        width: 200,
                        heightImage: 140,
                        imageProvider: AssetImage('assets/mockup.png'),
                        tags: [_tag('Category', () {}), _tag('Product', () {})],
                        footer: _footer(),
                      ),
                      const SizedBox(width: 12),
                      FillImageCard(
                        width: 200,
                        heightImage: 140,
                        imageProvider: AssetImage('assets/mockup.png'),
                        footer: _footer(),
                      ),
                    ],
                  ),
                ),
                const SizedBox(height: 12),
                SingleChildScrollView(
                  scrollDirection: Axis.horizontal,
                  child: Row(
                    children: [
                      TransparentImageCard(
                        width: 200,
                        imageProvider: AssetImage('assets/mockup.png'),
                        tags: [
                          _tag('Product', () {}),
                        ],
                        title: _title(color: Colors.white),
                        description: _content(color: Colors.white),
                      ),
                      const SizedBox(width: 12),
                      TransparentImageCard(
                        width: 200,
                        imageProvider: AssetImage('assets/mockup.png'),
                        tags: [
                          _tag('Product', () {}),
                          _tag('Gift', () {}),
                        ],
                        title: _title(color: Colors.white),
                        description: _content(color: Colors.white),
                        footer: _footer(color: Colors.white),
                      ),
                      const SizedBox(width: 12),
                      TransparentImageCard(
                        width: 200,
                        imageProvider: AssetImage('assets/mockup.png'),
                        tags: [
                          _tag('Product', () {}),
                          _tag('Gift', () {}),
                          _tag('Category', () {}),
                        ],
                        footer: _footer(color: Colors.white),
                      ),
                      const SizedBox(width: 12),
                      TransparentImageCard(
                        width: 200,
                        imageProvider: AssetImage('assets/mockup.png'),
                        title: _title(color: Colors.white),
                        description: _content(color: Colors.white),
                      ),
                    ],
                  ),
                ),
                const SizedBox(height: 12),
              ],
            )),
      ),
    );
  }

  Widget _title({Color? color}) {
    return Text(
      'Card title',
      style: TextStyle(fontSize: 16, fontWeight: FontWeight.w600, color: color),
    );
  }

  Widget _content({Color? color}) {
    return Text(
      'This a card description',
      style: TextStyle(color: color),
    );
  }

  Widget _footer({Color? color}) {
    return Row(
      children: [
        CircleAvatar(
          backgroundImage: AssetImage(
            'assets/avatar.png',
          ),
          radius: 12,
        ),
        const SizedBox(
          width: 4,
        ),
        Expanded(
            child: Text(
          'Super user',
          style: TextStyle(color: color),
        )),
        IconButton(onPressed: () {}, icon: Icon(Icons.share))
      ],
    );
  }

  Widget _tag(String tag, VoidCallback onPressed) {
    return InkWell(
      onTap: onPressed,
      child: Container(
        decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(6), color: Colors.green),
        padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 2),
        child: Text(
          tag,
          style: TextStyle(color: Colors.white),
        ),
      ),
    );
  }
}
